### PR TITLE
feature(sdk): Optimize unnecessary condition task when using User definable CEL Custom task 

### LINKED ISSUE #498: Support generate "custom task" defination in DSL
/kind feature

**Description:**
I think we need support to generate `custom task` definition to the yaml file in `dsl`, like this:
```
      - name: exception-handler
        taskRef:
          apiVersion: custom.tekton.dev/v1alpha1
          kind: Exception
          name: exception
        params:
        - name: pipelinerun_name
          value: guarded-pr
```

I think user will have such requirement, since the `tekton` introduced the conception of `custom task`.
with such ability we can easy to support `loop`, `exception handler` .etc which depends on controller of `custom task`.

**Additional information:**

https://docs.google.com/document/d/14PEuKY6T1ku4uFZx12FLFEPNE8poTdZRlzj6RDT3YKk/edit

### PR DESCRIPTION
**Pull request by Tomcli**
**Which issue is resolved by this Pull Request:** 
Resolves #498

**Description of your changes:**
This PR removes the unnecessary condition task when the `==` and `!=` comparison is done between the CEL Custom task outputs because CEL already evaluated the condition for us. 

This PR also includes a development feature using [tektoncd/experimental CEL](https://github.com/tektoncd/experimental/tree/main/cel) for conditions. However, the experimental CEL is still evolving, so we are not exposing it to the end users until it's mature. KFP-Tekton SDK developer can update the `DISABLE_CEL_CONDITION ` variable in the compiler.py code to preview the implementation.

old Design docs https://docs.google.com/document/d/1aE9yRdY_q672kqO5d0N4y6ngtCR3YKDahHk6ejmpJJU/edit#heading=h.tvifzmf8mj1m

/cc @rafalbigaj @Udiknedormin 

**Environment tested:**

* Python Version (use `python --version`): 3.7
* Tekton Version (use `tkn version`): 0.21.0
* Kubernetes Version (use `kubectl version`): 1.18
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 1]
```diff
@@ -373,15 +383,26 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
+        else:
+          condition_task_ref = [{
+              'name': template['metadata']['name'],
+              'params': [{
+                  'name': p['name'],
+                  'value': p.get('value', '')
+                } for p in template['spec'].get('params', [])
+              ],
+              'taskRef': _get_cel_condition_template()
+          }]
```
[the remark this answers is not in the document]
Comment (Tomcli) [reply sdk/python/kfp_tekton/compiler/compiler.py:405]: thanks @Udiknedormin

**Comment by Tomcli** [comment]
@googlebot I fixed it.

File: sdk/python/tests/compiler/testdata/conditions_and_loops.yaml [hunk 2]
```diff
@@ -15,103 +15,49 @@
...
+    tekton.dev/output_artifacts: '{"add-numbers": [{"key": "artifacts/$PIPELINERUN/add-numbers/Output.tgz",
+      "name": "add-numbers-Output", "path": "/tmp/outputs/Output/data"}], "print-number":
+      [{"key": "artifacts/$PIPELINERUN/print-number/Output.tgz", "name": "print-number-Output",
+      "path": "/tmp/outputs/Output/data"}], "produce-numbers": [{"key": "artifacts/$PIPELINERUN/produce-numbers/Output.tgz",
+      "name": "produce-numbers-Output", "path": "/tmp/outputs/Output/data"}]}'
+  name: conditions-and-loops
 spec:
   params:
-  - name: "n"
+  - name: n
```
Comment (drewbutlerbb4) [inline sdk/python/tests/compiler/testdata/conditions_and_loops.yaml:39]: The yaml validator in the UI has issues with `n` being used as the name. Can we change `n` to `num` or something else in  `sdk/python/tests/compiler/testdata/conditions_and_loops.py`

File: sdk/python/tests/compiler/testdata/condition.yaml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2020 kubeflow.org
+# Copyright 2021 kubeflow.org
```
Comment (drewbutlerbb4) [inline sdk/python/tests/compiler/testdata/condition.yaml:1]: The UI has issues displaying Conditions after this update. I can fix this in a subsequent PR.

File: sdk/python/tests/compiler/testdata/conditions_with_global_params.yaml [hunk 4]
```diff
@@ -15,107 +15,58 @@
...
+      "parent_task": "add-numbers"}]}'
+    tekton.dev/output_artifacts: '{"add-numbers": [{"key": "artifacts/$PIPELINERUN/add-numbers/Output.tgz",
+      "name": "add-numbers-Output", "path": "/tmp/outputs/Output/data"}], "print-number":
+      [{"key": "artifacts/$PIPELINERUN/print-number/Output.tgz", "name": "print-number-Output",
+      "path": "/tmp/outputs/Output/data"}]}'
+  name: conditions-with-global-params
 spec:
   params:
-  - name: "n"
+  - name: n
```
Comment (drewbutlerbb4) [inline sdk/python/tests/compiler/testdata/conditions_with_global_params.yaml:38]: > The yaml validator in the UI has issues with n being used as the name. Can we change n to num or something else in sdk/python/tests/compiler/testdata/conditions_and_loops.py

The same applies here.

Comment (drewbutlerbb4) [reply sdk/python/tests/compiler/testdata/conditions_with_global_params.yaml:38]: Needs to be fixed in `sdk/python/tests/compiler/testdata/conditions_with_global_params.py`

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 5]
```diff
@@ -505,26 +519,53 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
+              'taskSpec' if DISABLE_CEL_CONDITION else 'taskRef': condition_task_spec
+          }]
+          condition_refs[template['metadata']['name']] = [
+              {
+                'input': '$(tasks.%s.results.status)' % template['metadata']['name'],
+                'operator': 'in',
+                'values': ['true']
+              }
+            ]
+          # Don't use additional task if it's only doing literal string == and !=
```
Comment (Udiknedormin) [inline sdk/python/kfp_tekton/compiler/compiler.py:552]: That's actually misleading, as it breaks the 1:1 relationship between DSL and Tekton entities. `dsl.Condition` is expected to generate a single task or custom task, if it generates it only for some operators, tracking becomes very unobvious.

```python
with dsl.Condition(op.['results_len'] > 0): # custom task is generated
  ...


with dsl.Condition(op.output['status'] == "succeeded"):  # custom task is not generated
  ...
```

File: samples/flip-coin-custom-task/condition.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        with dsl.Condition(random_num_head.output <= 5):
+            print_op('heads and %s <= 5!' % random_num_head.output)
+
+    with dsl.Condition(cel_condition.output != 'true'):
+        random_num_tail = random_num_op(10, 19)
+        with dsl.Condition(random_num_tail.output > 15):
+            inner_task = print_op('tails and %s > 15!' % random_num_tail.output)
+        with dsl.Condition(random_num_tail.output <= 15):
+            print_op('tails and %s <= 15!' % random_num_tail.output)
+    random_num_head2 = random_num_op(0, 9).after(random_num_head).after(random_num_tail).after(inner_task)
```
[the remark this answers is not in the document]
Comment (Tomcli) [reply samples/flip-coin-custom-task/condition.py]: done

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 7]
```diff
@@ -505,26 +519,53 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
+          condition_operator = condition_params[2]
+          condition_operand1 = condition_params[0]
+          condition_operand2 = condition_params[1]
+          if (condition_operator.get('value', '') == '==' or condition_operator.get('value', '') == '!=') and not DISABLE_CEL_CONDITION:
+            map_cel_vars = lambda a: '$(tasks.%s.results.%s)' % (sanitize_k8s_name(a['value'].split('.')[-1]),
+              sanitize_k8s_name(a['output_name'])) if a.get('type', '') == dsl.PipelineParam else a.get('value', '')
+            condition_refs[template['metadata']['name']] = [
+                {
+                  'input': map_cel_vars(condition_operand1),
+                  'operator': 'in' if condition_operator.get('value', '') == '==' else 'notin',
```
[the remark this answers is not in the document]
Comment (Tomcli) [reply sdk/python/kfp_tekton/compiler/compiler.py]: done

File: sdk/python/kfp_tekton/compiler/compiler.py (hunk 5 above)
Comment (Tomcli) [reply sdk/python/kfp_tekton/compiler/compiler.py:551]: done

**Review by Udiknedormin — APPROVED** [review]

**Comment by Tomcli** [comment]
@drewbutlerbb4 Can you lgtm this PR? Looks like Michalina is not in the Kubeflow org so she can't lgtm it.

**Pull request merged**